### PR TITLE
Reduce string allocations on decoding values parsed from strings

### PR DESCRIPTION
**Pull request by tsholmes**
This reduces string allocations where we are decoding non-string values by parsing them out of a string.

This takes advantage of the fact that the parsing functions do not retain a reference to the passed in string, so it is safe to use a string created from the mutable buffer that will be overwritten after the parse.

### CODE REVIEW

**Comment by achille-roussel** [comment]
That's a good initiative, unfortunately I think it may lead to unexpected behaviors.

https://godoc.org/github.com/segmentio/objconv#Parser
>  // ParseBytes parses a byte array value.
>  //
>  // The returned byte slice is expected to be pointing at an internal memory
>  // buffer, the decoder will make a copy of the value. This design allows more
>  // memory allocation optimizations.
>  ParseBytes() ([]byte, error)
(same goes for ParseString).

Another option I was thinking we could take is allocate 1KB buffers and takes slices of them to generate the small strings efficiently. But that could result in retaining lots of large buffers if only a small portion of the small strings are referenced by the application, leading to much higher memory utilization.

It's a tradeoff that may be acceptable in some cases, so maybe we could make this an opt-in feature.

**Comment by tsholmes** [comment]
The strings being created here will only be retained for the duration of the type-specific parse calls through right? (e.g. `time.ParseDuration`). For these cases the string created only exists for the duration of that parse, and the reference is released before the next read from the objconv Parser, so it won't cause any retention of the backing byte slice later on.

**Comment by tsholmes** [comment]
Instead of having to cover the different error types that would need to be deep copied, I updated it to just call the parse method again with a "safe" string on error. does this seem like a good solution? or would you rather I had an explicit deep copy of the returned errors?

File: decode.go [hunk 1]
```diff
@@ -1555,3 +1580,21 @@ func makeDecodePtrFunc(t reflect.Type, opts decodeFuncOpts) decodeFunc {
...
+// - the bytes pointed to by b will not be modified while the returned string exists
+// - the returned string will not be stored past the lifetime of b
+// if the method being called with an unsafe returns an error, that error may
+// contain a reference to unsafe string. if the method produces consistent results,
+// calling it again with a safe string should return an error that can be safely
+// returned to the caller.
+func unsafeString(b []byte) string {
+	if len(b) == 0 {
+		return ""
+	}
```
Comment (achille-roussel) [inline decode.go]: If you're looking at even more performance you can use this implementation:
```go
return *(*string)(unsafe.Pointer(&b))
```
last I checked functions containing a `if` statement don't get inlined in Go.

File: decode.go [hunk 2]
```diff
@@ -174,7 +175,10 @@ func (d Decoder) decodeIntFromType(t Type, to reflect.Value) (err error) {
 			return
 		}
 
-		i, err = strconv.ParseInt(string(b), 10, 64)
+		i, err = strconv.ParseInt(unsafeString(b), 10, 64)
+		if err != nil {
+			_, err = strconv.ParseInt(string(b), 10, 64)
```
Comment (achille-roussel) [inline decode.go]: Maybe leave a comment here explaining why we have to redo the operation to generate the error?

**Review by achille-roussel — APPROVED** [review]
Nice hack 👍

**Pull request merged**